### ISSUE TITLE
Move extensibility to use-site

### ISSUE THREAD

**Issue by robinheghan** [gren-lang/compiler#240]
Copying this idea from elm slack

Haha yeah, I didn't really link the two, but went I went to post Zokka to the Elm Discourse, I figured there wasn't really much point in cloaking who I was. RE Roc, I agree I've been meaning to try it out. There's one small UX thing about composability that I wish Roc did differently (essentially I don't think that type synonyms themselves should express whether they are extensible, but the extensibility should be delayed to the call site, so that e.g. in hypothetical Elm syntax type MyRecord = {a : Int, b : String} could be used at the callsite as myFunctionWithExtensibleRecord :: a ++ MyRecord -> Int, which would also allow type synonyms to be nicely merged together (e.g. type MyBigRecord = MyRecord ++ MyOtherRecord) without needing to decide beforehand which type synonyms are meant to be extensible and which aren't, since that's usually practically ends up being more relevant for callsites

**Comment by boxed** [comment]
Minor quibble: I find `++` to be a weird syntax. That's the append operator right? Append isn't commutative.

I like this a lot. It's a lot nicer than going around copy-pasting `{a | foo: int, bar: baz}` in many places.

**Comment by blaix** [comment]
> Minor quibble: I find `++` to be a weird syntax. That's the append operator right? Append isn't commutative.

What about javsacript's spread operator?

Something like: `{ a : String, ...OtherRecord }`

Could be used in pattern matching too: `{ a, ...everythingElse }`, `[ "blog", "post", ...path ]` :)

> I like this a lot.

Me too!

My gut reaction was wondering if this could be abused. In other words, is keeping it explicit per record protecting us from ourselves? But I don't think so.

**Comment by boxed** [comment]
@blaix it sounds like you are talking's about two different things.

**Comment by blaix** [comment]
Very likely! I'm easily confused. Happy to delete if I'm talking nonsense.

**Comment by robinheghan** [comment]
If I'm understanding @blaix correctly, you'd be able to extend a record like ``` { ...Record1, ...Record2 }``` in addition to doing it inline, which his first example is doing. Is that right, @blaix ?

For records it works fine. For Custom Types/Tags, I guess it could work like ```| Tag1 | Tag2 | ...TagAlias``` and ``` | ...Alias1 | ...Alias2```?

I do like the idea of matching this up with pattern matching, but I think that's a discussion we can have on Zulip.

**Comment by blaix** [comment]
Right, I think I focused too much on the `A = B ++ C` part of the example. I see now where I misunderstood the function signature part of the suggestion. I guess that would look something like `greet : { ...Named } -> String`. Or maybe as part of the parameter definition: `greet { name, ... } =` (like the hypothetical pattern matching)
